### DISCUSSION TITLE
How to properly implement the check whether assets are downloaded or not?

### DISCUSSION THREAD

**Discussion by ComradeRevo** [jelte/be.khepri.play.assetdelivery.addressables#23] [General]
I have four groups in my project: Campaign, CustomGame, Backgrounds, Preview. All assets is Sprites with corresponding labels. before implementing AssetDelivery.Addressables, I used the following code:

```
private IEnumerator CheckAndDownloadRemoteAssetsAsync()
    {        
        List<object> labels = new List<object>();

        labels.Add("Campaign");
        labels.Add("CustomGame");
        labels.Add("Background");
        labels.Add("Preview");

        var isMenuActive = false;

        foreach (var label in labels)
        {
            //var downloadSize = Addressables.GetDownloadSizeAsync(label);
            var downloadSize = Khepri.AssetDelivery.AddressablesAssetDelivery.GetDownloadSizeAsync(label);

            yield return downloadSize;

            while (!downloadSize.IsDone)
            {
                yield return null;
            }

            if (downloadSize.Result > 0)
            {
                if (!isMenuActive)
                {
                    GUIManager.Instance.ShowMenu("ContentLoadingMenu");
                    isMenuActive = true;
                }

                var downloadDependencies = Addressables.DownloadDependenciesAsync(label);

                yield return downloadDependencies;

            }

            Debug.LogFormat("CheckAndDownloadRemoteAssetsAsync: GetDownloadSizeAsync({0}) = {1}", label, downloadSize.Result);
        }

        if (isMenuActive)
        {
            GUIManager.Instance.HideMenu("ContentLoadingMenu");
        }
    }
```

I didn't understand addressable too well, but everything seemed to work with Remote Group. During the first installation of the application, a check was performed, if assets have not yet been downloaded, then a menu is displayed and the download is performed.

Now Remote Group I change to Local, Bundle Mode = Pack Together, Asset Provider = Assets from Bundles Provider, Asset Bundle Provider = Asset Pack Bundle Proveder. It seems the check must be performed with the help of  Khepri.AssetDelivery.AddressablesAssetDelivery.GetDownloadSizeAsync(label)

Is it correct to change the assets size check, or do I need to perform the check in a different way?
Is it possible to create a development build to test the operation of AssetDelivery.Addressables without creating Android App Bundle for internal testing?

**Comment by jelte** [comment]
1.  Is it correct to change the assets size check, or do I need to perform the check in a different way?
Yes, that's correct although I would suggest you add both values instead so you can still support remote bundles.
```csharp
var downloadSize = Addressables.GetDownloadSizeAsync(labels) + Khepri.AssetDelivery.AddressablesAssetDelivery.GetDownloadSizeAsync(labels);
```
Note: You can pass the list of labels, you don't need to pass them one by one. The same is possible for ` Addressables.DownloadDependenciesAsync`.

2.  Is it possible to create a development build to test the operation of AssetDelivery.Addressables without creating Android App Bundle for internal testing?
No, there is no way to upload AssetPacks separately to the Play Store. But you can upload to a test track (Internal Testing) or use "Internal app sharing".

**Reply by ComradeRevo** [reply]
Thanks!
